### ISSUE TITLE
Parallel run gets finished status even when one task fails

### ISSUE THREAD

**Issue by nhammond** [StanfordBioinformatics/loom#451]
```json
{
    "status": "Finished",
    "is_leaf": true,
    "tasks": [
        {
            "uuid": "59090bc2-9427-46a2-85d1-5f53f5254196",
            "url": "https://10.142.6.5:443/api/tasks/59090bc2-9427-46a2-85d1-5f53f5254196/",
            "status": "Finished",
            "datetime_finished": "2017-08-29T15:29:46.649867Z",
            "datetime_created": "2017-08-29T15:08:42.919223Z"
        },
        {
            "uuid": "f2fb7465-238f-4861-8e03-a59f829f5b6d",
            "url": "https://10.142.6.5:443/api/tasks/f2fb7465-238f-4861-8e03-a59f829f5b6d/",
            "status": "Finished",
            "datetime_finished": "2017-08-29T16:11:15.826314Z",
            "datetime_created": "2017-08-29T15:43:50.776111Z"
        },
        {
            "uuid": "0de82d0d-9628-4486-a03a-e6415aa9c7f0",
            "url": "https://10.142.6.5:443/api/tasks/0de82d0d-9628-4486-a03a-e6415aa9c7f0/",
            "status": "Failed",
            "datetime_finished": null,
            "datetime_created": "2017-08-29T19:17:55.176730Z"
        }
    ],
    "uuid": "c9a08a18-2c2b-4301-b089-4c1d85d340e8",
    "status_is_killed": false,
    "url": "https://10.142.6.5:443/api/runs/c9a08a18-2c2b-4301-b089-4c1d85d340e8/",
    "outputs": [
        {
            "data": {
                "uuid": "f8191216-09b3-4293-a506-90f389dc4040",
                "url": "https://10.142.6.5:443/api/data-nodes/f8191216-09b3-4293-a506-90f389dc4040/"
            },
            "type": "file",
            "mode": "no_scatter",
            "channel": "qualityScoreDistrbution_chart_output",
            "source": {
                "filename": "{{sampleid}}.QualityScoreDistribution.histogram.pdf"
            }
        },
        {
            "data": {
                "uuid": "600f9d6c-e796-4433-a87f-46076f4645cf",
                "url": "https://10.142.6.5:443/api/data-nodes/600f9d6c-e796-4433-a87f-46076f4645cf/"
            },
            "type": "file",
            "mode": "no_scatter",
            "channel": "qualityScoreDistribution_output_txt",
            "source": {
                "filename": "{{sampleid}}.qualityscoredistribution.output.txt"
            }
        }
    ],
    "datetime_created": "2017-08-29T05:48:52.618553Z",
    "status_is_waiting": false,
    "status_is_running": true,
    "inputs": [
        {
            "data": {
                "uuid": "f8bde733-6d30-4874-b05e-fd82711e431e",
                "url": "https://10.142.6.5:443/api/data-nodes/f8bde733-6d30-4874-b05e-fd82711e431e/"
            },
            "type": "file",
            "mode": "no_gather",
            "channel": "recalibrated_bam",
            "group": 1
        },
        {
            "data": {
                "uuid": "17108b65-8bd6-4ad3-93c1-9414a15e2674",
                "url": "https://10.142.6.5:443/api/data-nodes/17108b65-8bd6-4ad3-93c1-9414a15e2674/"
            },
            "type": "file",
            "mode": "no_gather",
            "channel": "recalibrated_bam_bai",
            "group": 1
        },
        {
            "data": {
                "uuid": "bfd09db8-5256-41fb-806f-5e08b1a2ffcf",
                "url": "https://10.142.6.5:443/api/data-nodes/bfd09db8-5256-41fb-806f-5e08b1a2ffcf/"
            },
            "type": "file",
            "mode": "no_gather",
            "channel": "reference",
            "group": 0
        },
        {
            "data": {
                "uuid": "7c99ca84-3f9d-492e-8736-863ab5898d22",
                "url": "https://10.142.6.5:443/api/data-nodes/7c99ca84-3f9d-492e-8736-863ab5898d22/"
            },
            "type": "file",
            "mode": "no_gather",
            "channel": "reference_dict",
            "group": 0
        },
        {
            "data": {
                "uuid": "f4ad5c13-7259-481f-bddf-f00e3f5c836e",
                "url": "https://10.142.6.5:443/api/data-nodes/f4ad5c13-7259-481f-bddf-f00e3f5c836e/"
            },
            "type": "file",
            "mode": "no_gather",
            "channel": "reference_index_fai",
            "group": 0
        },
        {
            "data": {
                "uuid": "75042339-fefa-4ca3-b5d9-fcacc6a98f7b",
                "url": "https://10.142.6.5:443/api/data-nodes/75042339-fefa-4ca3-b5d9-fcacc6a98f7b/"
            },
            "type": "string",
            "mode": "no_gather",
            "channel": "sampleid",
            "group": 1
        }
    ],
    "command": "/usr/bin/java -Xms6g -Xmx8g -jar /usr/gitc/picard.jar  QualityScoreDistribution ALIGNED_READS_ONLY=true R={{reference}} INPUT={{recalibrated_bam}} OUTPUT={{qualityScoreDistribution_output_txt}} ASSUME_SORTED=True CHART={{qualityScoreDistrbution_chart_output}} PF_READS_ONLY=True",
    "resources": {
        "cores": "1",
        "memory": "16"
    },
    "template": {
        "uuid": "3c87e800-f990-4038-987c-2e9f5edfee7c",
        "url": "https://10.142.6.5:443/api/templates/3c87e800-f990-4038-987c-2e9f5edfee7c/",
        "name": "qualityScoreDistribution_Picard",
        "md5": null,
        "datetime_created": "2017-08-25T20:20:56.325607Z",
        "is_leaf": true
    },
    "interpreter": "/bin/bash -euo pipefail",
    "postprocessing_status": "complete",
    "status_is_failed": false,
    "status_is_finished": true,
    "environment": {
        "docker_image": "vandhanak/broadcloudgatk36gatk37-cgs:v1"
    },
    "name": "qualityScoreDistribution_Picard"
}
```

**Comment by nhammond** [comment]
When a task under a parallel run finishes, the run simply checks if all existing tasks are finished, and if so the run is also flagged as finished. But there may be tasks that have not yet been created, and the run does not take these into account. 

This bad state, with a finished run that contains not-yet-completed tasks, is reproducible with the workflow below.

```
name: sleep
inputs:
- type: string
  channel: delay
  data:
    contents: [1,2,60]
outputs:
- type: string
  channel: out2
steps:
- name: delay_step
  command: sleep {{delay}}; echo {{delay}}
  environment:
    docker_image: ubuntu
  resources:
    cores: '1'
    memory: '1'
  inputs:
  - type: string
    channel: delay
  outputs:
  - type: string
    channel: out1
    source:
      stream: stdout
- name: echo_step
  command: echo {{ out1 }}
  environment:
    docker_image: ubuntu
  resources:
    cores: '1'
    memory: '1'
  inputs:
  - type: string
    channel: out1
  outputs:
  - type: string
    channel: out2
    source:
      stream: stdout
```

**Comment by nhammond** [comment]
Fixed in 90c3ecdf532e39aa2016afe0a5ebd12d2d7a2fde
